src/framework/constraints.ml

```diff
@@ -991,6 +991,39 @@ struct
     if D.is_bot d then raise Deadcode else d
 end
 
+module DeadBranchLifter (S: Spec): Spec =
+struct
+  include S
+
+  let name () = "DeadBranch (" ^ name () ^ ")"
+
+  module Locmap = Deadcode.Locmap
+
+  let dead_branches = function true -> Deadcode.dead_branches_then | false -> Deadcode.dead_branches_else
+
+  let locmap_modify_def d k f h =
+    if Locmap.mem h k then
+      Locmap.replace h k (f (Locmap.find h k))
+    else
+      Locmap.add h k d
+
+  let branch ctx exp tv =
+    if !GU.in_verifying_stage && get_bool "dbg.print_dead_code" then (
+      Locmap.replace Deadcode.dead_branches_cond !Tracing.current_loc exp;
+      try
+        let r = branch ctx exp tv in
+        (* branch is live *)
+        Locmap.replace (dead_branches tv) !Tracing.current_loc false;
+        r
+      with Deadcode ->
+        (* branch is dead *)
+        locmap_modify_def true !Tracing.current_loc (fun x -> x) (dead_branches tv);
```

Comment:
what is the deal with `fun x -> x` here? This is the only place where `locmap_modify_def` is called, why does it need such a weird argument?

Reply:
This is related to the multiple contexts thing above. When a branch is found to be live, then it's deadness is always overwritten to `false`. When a branch is found to be dead here, then it's set to dead (`true`) if it didn't exist in the map before, otherwise the identity function keeps it as it was. If it has always been dead (`true`) then it stays that way, if in some case it was made live (`false`), we don't want to overwrite it, but preserve the liveness.

Reply:
Yes, I agree this is what we want to do. However, this `fun x-> x` somehow obfuscates that more than it helps.

Reply:
I added comments explaining this and replaced `locmap_modify_def` with the equivalent `Locmap.modify_def`.